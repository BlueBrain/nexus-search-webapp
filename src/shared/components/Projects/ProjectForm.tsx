import * as React from 'react';
import { Collapse, Form, Icon, Input, Button, Spin, Modal } from 'antd';
import { WrappedFormUtils } from 'antd/lib/form/Form';

/**
 * Custom from controls, based on:
 * https://ant.design/components/form/#components-form-demo-customized-form-controls
 */
export interface PrefixMappingGroupInputState {
  prefix: string;
  namespace: string;
}
export interface PrefixMappingGroupInputProps {
  groupId: string | number;
  value: any;
  onChange(state: PrefixMappingGroupInputState): void;
}
class PrefixMappingGroupInput extends React.Component<
  PrefixMappingGroupInputProps,
  PrefixMappingGroupInputState
> {
  // Static method called by Form component
  static getDerivedStateFromProps(nextProps: PrefixMappingGroupInputProps) {
    if ('value' in nextProps) {
      return {
        ...(nextProps.value || {}),
      };
    }
    return null;
  }

  constructor(props: PrefixMappingGroupInputProps) {
    super(props);
    const value = props.value || {};
    this.state = {
      prefix: value.prefix || '',
      namespace: value.namespace || '',
    };
  }

  handlePrefixChange = (e: React.FormEvent<HTMLInputElement>) => {
    const prefix = e.currentTarget.value;
    this.setState({
      prefix,
    });
    this.props.onChange({ ...this.state, prefix });
  };

  handleNamespaceChange = (e: React.FormEvent<HTMLInputElement>) => {
    const namespace = e.currentTarget.value;
    this.setState({
      namespace,
    });
    this.props.onChange({ ...this.state, namespace });
  };

  render() {
    return (
      <Input.Group
        compact
        style={
          {
            width: 'calc(100% - 22px)',
            marginRight: 8,
          } /* icon is 14px + 8px margin-right = 22px*/
        }
      >
        <Input
          style={{ width: '20%' }}
          name={`prefix[${this.props.groupId}]`}
          placeholder="prefix"
          onChange={this.handlePrefixChange}
          value={this.state.prefix}
        />
        <Input
          style={{ width: '80%' }}
          name={`namespace[${this.props.groupId}]`}
          placeholder="namespace"
          onChange={this.handleNamespaceChange}
          value={this.state.namespace}
        />
      </Input.Group>
    );
  }
}

export interface ProjectFormProps {
  form: WrappedFormUtils;
  project?: {
    label: string;
    rev: number;
    description?: string;
    base?: string;
    vocab?: string;
    apiMappings?: PrefixMappingGroupInputState[];
  };
  busy?: boolean;
  onSubmit?(project: ProjectFormProps['project']): any;
  onDeprecate?(): any;
  onMakePublic?(): any;
  mode?: 'create' | 'edit';
}

/**
 * Adaptation of the following example:
 * based on: https://ant.design/components/form/#components-form-demo-dynamic-form-item
 */
const ProjectForm: React.FunctionComponent<ProjectFormProps> = ({
  form,
  project,
  busy = false,
  onSubmit = () => {},
  onDeprecate = () => {},
  onMakePublic = () => {},
  mode = 'create',
}) => {
  // logic for generating dynamic prefix mapping fields in form
  const currentId =
    project && project.apiMappings ? project.apiMappings.length : 0;
  const activeKeys = [...Array(currentId + 1).keys()].slice(1);
  const [prefixMappingKeys, setPrefixMappingKeys] = React.useState({
    currentId,
    activeKeys,
  });

  const { getFieldDecorator, getFieldValue } = form;
  const formItemLayout = {
    labelCol: {
      xs: { span: 24 },
      sm: { span: 6 },
    },
    wrapperCol: {
      xs: { span: 24 },
      sm: { span: 18 },
    },
  };
  const formItemLayoutWithOutLabel = {
    wrapperCol: {
      xs: { span: 24, offset: 0 },
      sm: { span: 18, offset: 6 },
    },
  };

  const add = (k: any) => {
    const { currentId, activeKeys } = prefixMappingKeys;
    const newId: number = currentId + 1;
    // @ts-ignore
    setPrefixMappingKeys({
      currentId: newId,
      activeKeys: [...activeKeys, newId],
    });
  };

  const remove = (k: any) => {
    const { activeKeys } = prefixMappingKeys;
    setPrefixMappingKeys({
      ...prefixMappingKeys,
      activeKeys: activeKeys.filter((key: any) => key !== k),
    });
  };

  const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
    form.validateFields((err, values) => {
      if (!err) {
        onSubmit({
          ...values,
          apiMappings:
            (values.apiMappings &&
              values.apiMappings.filter((p: any) => !!p)) ||
            [],
        });
      }
    });
  };

  const checkPrefix = (
    rule: {},
    value: PrefixMappingGroupInputState,
    callback: (message?: string) => void
  ) => {
    if (value.prefix && value.namespace) {
      callback();
      return;
    }
    callback('You need to specify both prefix and namespace');
  };

  const confirmDeprecate = () => {
    Modal.confirm({
      title: 'Deprecate Project',
      content: 'Are you sure?',
      onOk: onDeprecate,
    });
  };

  const confirmMakePublic = () => {
    Modal.confirm({
      title: 'Make Project Public',
      content: 'Are you sure?',
      onOk: onMakePublic,
    });
  };

  // Dynamic form fields
  const apiMappingsItems = prefixMappingKeys.activeKeys.map(
    (key: number, index: number) => (
      <Form.Item
        label={index === 0 ? 'API Mappings' : ''}
        {...(index === 0 ? formItemLayout : formItemLayoutWithOutLabel)}
        key={key}
      >
        {getFieldDecorator(`apiMappings[${key - 1}]`, {
          initialValue: {
            prefix:
              (project &&
                project.apiMappings &&
                project.apiMappings[key - 1] &&
                project.apiMappings[key - 1].prefix) ||
              '',
            namespace:
              (project &&
                project.apiMappings &&
                project.apiMappings[key - 1] &&
                project.apiMappings[key - 1].namespace) ||
              '',
          },
          rules: [{ validator: checkPrefix, required: true }],
        })(
          // @ts-ignore
          <PrefixMappingGroupInput groupId={`${key}`} />
        )}
        {prefixMappingKeys.activeKeys.length > 0 ? (
          <Icon
            className="dynamic-delete-button"
            type="minus-circle-o"
            onClick={() => remove(key)}
          />
        ) : null}
      </Form.Item>
    )
  );
  return (
    <Spin
      spinning={busy}
      tip="Please be patient while the project is scaffolded."
    >
      <Form onSubmit={handleSubmit}>
        <Form.Item label="Label" {...formItemLayout}>
          {getFieldDecorator('label', {
            initialValue: project ? project.label : '',
            rules: [
              {
                required: true,
                whitespace: true,
                pattern: /^\S+$/g,
                message: 'Label must be a phrase without spaces',
              },
            ],
          })(<Input placeholder="Label" disabled={mode === 'edit'} />)}
        </Form.Item>
        <Form.Item label="Description" {...formItemLayout}>
          {getFieldDecorator('description', {
            initialValue: project ? project.description : '',
            rules: [{ required: false }],
          })(<Input placeholder="Description" />)}
        </Form.Item>
        <Form.Item {...formItemLayoutWithOutLabel}>
          <Collapse>
            <Collapse.Panel header="Advanced settings" key="1">
              <Form.Item label="Base" {...formItemLayout}>
                {getFieldDecorator('base', {
                  initialValue: project ? project.base : '',
                  rules: [{ required: false }],
                })(<Input placeholder="Base" />)}
              </Form.Item>
              <Form.Item label="Vocab" {...formItemLayout}>
                {getFieldDecorator('vocab', {
                  initialValue: project ? project.vocab : '',
                  rules: [{ required: false }],
                })(<Input placeholder="Vocab" />)}
              </Form.Item>
              {apiMappingsItems}
              <Form.Item {...formItemLayoutWithOutLabel}>
                <Button type="dashed" onClick={add} style={{ width: '60%' }}>
                  <Icon type="plus" /> Add API mapping
                </Button>
              </Form.Item>
            </Collapse.Panel>
          </Collapse>
        </Form.Item>
        <Form.Item {...formItemLayoutWithOutLabel}>
          <Button
            type="primary"
            htmlType="submit"
            className="login-form-button"
          >
            Save
          </Button>
          {mode === 'edit' && (
            <>
              <Button
                type="danger"
                onClick={confirmDeprecate}
                style={{ float: 'right' }}
              >
                Deprecate
              </Button>
              <Button onClick={confirmMakePublic} style={{ float: 'right' }}>
                <Icon type="global" />
                Make public
              </Button>
            </>
          )}
        </Form.Item>
      </Form>
    </Spin>
  );
};

export default Form.create()(ProjectForm);
