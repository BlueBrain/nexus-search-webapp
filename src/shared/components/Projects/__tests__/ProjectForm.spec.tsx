import * as React from 'react';
import { mount } from 'enzyme';

// Need to mock Date.now() because it is used down the line
// to generate a per-element class, which breaks snapshot testing.
// Must be done before importing the React component.
jest.spyOn(Date, 'now').mockImplementation(() => 1234567890123);

import ProjectForm from '../ProjectForm';

const formComponent = <ProjectForm />;
const wrapper = mount(formComponent);

describe('Project Form component', () => {
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

// Restoring Date.now() to the real thing, so as not to break other tests.
jest.restoreAllMocks();
