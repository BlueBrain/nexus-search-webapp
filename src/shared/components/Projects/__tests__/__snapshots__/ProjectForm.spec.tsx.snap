// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Form component should render correctly 1`] = `
<Form(ProjectForm)>
  <ProjectForm
    form={
      Object {
        "getFieldDecorator": [Function],
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldProps": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "isSubmitting": [Function],
        "resetFields": [Function],
        "setFields": [Function],
        "setFieldsInitialValue": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
        "validateFieldsAndScroll": [Function],
      }
    }
  >
    <Spin
      size="default"
      spinning={false}
      tip="Please be patient while the project is scaffolded."
      wrapperClassName=""
    >
      <div
        className="ant-spin-nested-loading"
      >
        <div
          className="ant-spin-container"
          key="container"
        >
          <Form
            colon={true}
            hideRequiredMark={false}
            layout="horizontal"
            onSubmit={[Function]}
          >
            <form
              className="ant-form ant-form-horizontal"
              onSubmit={[Function]}
            >
              <FormItem
                hasFeedback={false}
                label="Label"
                labelCol={
                  Object {
                    "sm": Object {
                      "span": 6,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
                wrapperCol={
                  Object {
                    "sm": Object {
                      "span": 18,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-label"
                      sm={
                        Object {
                          "span": 6,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-label ant-col-xs-24 ant-col-sm-6"
                      >
                        <label
                          className="ant-form-item-required"
                          htmlFor="label"
                          onClick={[Function]}
                          title="Label"
                        >
                          Label
                        </label>
                      </div>
                    </Col>
                    <Col
                      className="ant-form-item-control-wrapper"
                      sm={
                        Object {
                          "span": 18,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-18"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Input
                              data-__field={
                                Object {
                                  "name": "label",
                                }
                              }
                              data-__meta={
                                Object {
                                  "initialValue": "",
                                  "name": "label",
                                  "originalProps": Object {
                                    "disabled": false,
                                    "placeholder": "Label",
                                    "type": "text",
                                  },
                                  "ref": null,
                                  "rules": Array [
                                    Object {
                                      "message": "Label must be a phrase without spaces",
                                      "pattern": /\\^\\\\S\\+\\$/g,
                                      "required": true,
                                      "whitespace": true,
                                    },
                                  ],
                                  "trigger": "onChange",
                                  "validate": Array [
                                    Object {
                                      "rules": Array [
                                        Object {
                                          "message": "Label must be a phrase without spaces",
                                          "pattern": /\\^\\\\S\\+\\$/g,
                                          "required": true,
                                          "whitespace": true,
                                        },
                                      ],
                                      "trigger": Array [
                                        "onChange",
                                      ],
                                    },
                                  ],
                                  "valuePropName": "value",
                                }
                              }
                              disabled={false}
                              id="label"
                              onChange={[Function]}
                              placeholder="Label"
                              type="text"
                              value=""
                            >
                              <input
                                className="ant-input"
                                data-__field={
                                  Object {
                                    "name": "label",
                                  }
                                }
                                data-__meta={
                                  Object {
                                    "initialValue": "",
                                    "name": "label",
                                    "originalProps": Object {
                                      "disabled": false,
                                      "placeholder": "Label",
                                      "type": "text",
                                    },
                                    "ref": null,
                                    "rules": Array [
                                      Object {
                                        "message": "Label must be a phrase without spaces",
                                        "pattern": /\\^\\\\S\\+\\$/g,
                                        "required": true,
                                        "whitespace": true,
                                      },
                                    ],
                                    "trigger": "onChange",
                                    "validate": Array [
                                      Object {
                                        "rules": Array [
                                          Object {
                                            "message": "Label must be a phrase without spaces",
                                            "pattern": /\\^\\\\S\\+\\$/g,
                                            "required": true,
                                            "whitespace": true,
                                          },
                                        ],
                                        "trigger": Array [
                                          "onChange",
                                        ],
                                      },
                                    ],
                                    "valuePropName": "value",
                                  }
                                }
                                disabled={false}
                                id="label"
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="Label"
                                type="text"
                                value=""
                              />
                            </Input>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                hasFeedback={false}
                label="Description"
                labelCol={
                  Object {
                    "sm": Object {
                      "span": 6,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
                wrapperCol={
                  Object {
                    "sm": Object {
                      "span": 18,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-label"
                      sm={
                        Object {
                          "span": 6,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-label ant-col-xs-24 ant-col-sm-6"
                      >
                        <label
                          className=""
                          htmlFor="description"
                          onClick={[Function]}
                          title="Description"
                        >
                          Description
                        </label>
                      </div>
                    </Col>
                    <Col
                      className="ant-form-item-control-wrapper"
                      sm={
                        Object {
                          "span": 18,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-18"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Input
                              data-__field={
                                Object {
                                  "name": "description",
                                }
                              }
                              data-__meta={
                                Object {
                                  "initialValue": "",
                                  "name": "description",
                                  "originalProps": Object {
                                    "disabled": false,
                                    "placeholder": "Description",
                                    "type": "text",
                                  },
                                  "ref": null,
                                  "rules": Array [
                                    Object {
                                      "required": false,
                                    },
                                  ],
                                  "trigger": "onChange",
                                  "validate": Array [
                                    Object {
                                      "rules": Array [
                                        Object {
                                          "required": false,
                                        },
                                      ],
                                      "trigger": Array [
                                        "onChange",
                                      ],
                                    },
                                  ],
                                  "valuePropName": "value",
                                }
                              }
                              disabled={false}
                              id="description"
                              onChange={[Function]}
                              placeholder="Description"
                              type="text"
                              value=""
                            >
                              <input
                                className="ant-input"
                                data-__field={
                                  Object {
                                    "name": "description",
                                  }
                                }
                                data-__meta={
                                  Object {
                                    "initialValue": "",
                                    "name": "description",
                                    "originalProps": Object {
                                      "disabled": false,
                                      "placeholder": "Description",
                                      "type": "text",
                                    },
                                    "ref": null,
                                    "rules": Array [
                                      Object {
                                        "required": false,
                                      },
                                    ],
                                    "trigger": "onChange",
                                    "validate": Array [
                                      Object {
                                        "rules": Array [
                                          Object {
                                            "required": false,
                                          },
                                        ],
                                        "trigger": Array [
                                          "onChange",
                                        ],
                                      },
                                    ],
                                    "valuePropName": "value",
                                  }
                                }
                                disabled={false}
                                id="description"
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="Description"
                                type="text"
                                value=""
                              />
                            </Input>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                hasFeedback={false}
                wrapperCol={
                  Object {
                    "sm": Object {
                      "offset": 6,
                      "span": 18,
                    },
                    "xs": Object {
                      "offset": 0,
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-control-wrapper"
                      sm={
                        Object {
                          "offset": 6,
                          "span": 18,
                        }
                      }
                      xs={
                        Object {
                          "offset": 0,
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-18 ant-col-sm-offset-6"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Collapse
                              bordered={true}
                              openAnimation={
                                Object {
                                  "appear": [Function],
                                  "enter": [Function],
                                  "leave": [Function],
                                }
                              }
                            >
                              <Collapse
                                accordion={false}
                                bordered={true}
                                className=""
                                destroyInactivePanel={false}
                                expandIcon={[Function]}
                                onChange={[Function]}
                                openAnimation={
                                  Object {
                                    "appear": [Function],
                                    "enter": [Function],
                                    "leave": [Function],
                                  }
                                }
                                prefixCls="ant-collapse"
                              >
                                <div
                                  className="ant-collapse"
                                  role={null}
                                >
                                  <CollapsePanel
                                    accordion={false}
                                    destroyInactivePanel={false}
                                    expandIcon={[Function]}
                                    header="Advanced settings"
                                    isActive={false}
                                    key=".$1"
                                    onItemClick={[Function]}
                                    openAnimation={
                                      Object {
                                        "appear": [Function],
                                        "enter": [Function],
                                        "leave": [Function],
                                      }
                                    }
                                    panelKey="1"
                                    prefixCls="ant-collapse"
                                  >
                                    <CollapsePanel
                                      accordion={false}
                                      className=""
                                      destroyInactivePanel={false}
                                      expandIcon={[Function]}
                                      forceRender={false}
                                      header="Advanced settings"
                                      headerClass=""
                                      isActive={false}
                                      onItemClick={[Function]}
                                      openAnimation={
                                        Object {
                                          "appear": [Function],
                                          "enter": [Function],
                                          "leave": [Function],
                                        }
                                      }
                                      panelKey="1"
                                      prefixCls="ant-collapse"
                                      showArrow={true}
                                    >
                                      <div
                                        className="ant-collapse-item"
                                      >
                                        <div
                                          aria-expanded="false"
                                          className="ant-collapse-header"
                                          onClick={[Function]}
                                          onKeyPress={[Function]}
                                          role="button"
                                          tabIndex={0}
                                        >
                                          <expandIcon
                                            accordion={false}
                                            className=""
                                            destroyInactivePanel={false}
                                            expandIcon={[Function]}
                                            forceRender={false}
                                            header="Advanced settings"
                                            headerClass=""
                                            isActive={false}
                                            onItemClick={[Function]}
                                            openAnimation={
                                              Object {
                                                "appear": [Function],
                                                "enter": [Function],
                                                "leave": [Function],
                                              }
                                            }
                                            panelKey="1"
                                            prefixCls="ant-collapse"
                                            showArrow={true}
                                          >
                                            <Icon
                                              className="ant-collapse-arrow"
                                              type="right"
                                            >
                                              <LocaleReceiver
                                                componentName="Icon"
                                              >
                                                <i
                                                  aria-label="icon: right"
                                                  className="anticon anticon-right ant-collapse-arrow"
                                                >
                                                  <IconReact
                                                    className=""
                                                    type="right-o"
                                                  >
                                                    <svg
                                                      aria-hidden="true"
                                                      className=""
                                                      data-icon="right"
                                                      fill="currentColor"
                                                      height="1em"
                                                      key="svg-right"
                                                      viewBox="64 64 896 896"
                                                      width="1em"
                                                    >
                                                      <path
                                                        d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                                                        key="svg-right-svg-0"
                                                      />
                                                    </svg>
                                                  </IconReact>
                                                </i>
                                              </LocaleReceiver>
                                            </Icon>
                                          </expandIcon>
                                          Advanced settings
                                        </div>
                                        <Animate
                                          animation={
                                            Object {
                                              "appear": [Function],
                                              "enter": [Function],
                                              "leave": [Function],
                                            }
                                          }
                                          component=""
                                          componentProps={Object {}}
                                          exclusive={true}
                                          onAppear={[Function]}
                                          onEnd={[Function]}
                                          onEnter={[Function]}
                                          onLeave={[Function]}
                                          showProp="isActive"
                                          transitionAppear={false}
                                          transitionEnter={true}
                                          transitionLeave={true}
                                        >
                                          <AnimateChild
                                            animation={
                                              Object {
                                                "appear": [Function],
                                                "enter": [Function],
                                                "leave": [Function],
                                              }
                                            }
                                            key="rc_animate_1234567890123"
                                            transitionAppear={false}
                                            transitionEnter={true}
                                            transitionLeave={true}
                                          >
                                            <PanelContent
                                              destroyInactivePanel={false}
                                              forceRender={false}
                                              isActive={false}
                                              key="rc_animate_1234567890123"
                                              prefixCls="ant-collapse"
                                              role={null}
                                            />
                                          </AnimateChild>
                                        </Animate>
                                      </div>
                                    </CollapsePanel>
                                  </CollapsePanel>
                                </div>
                              </Collapse>
                            </Collapse>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                hasFeedback={false}
                wrapperCol={
                  Object {
                    "sm": Object {
                      "offset": 6,
                      "span": 18,
                    },
                    "xs": Object {
                      "offset": 0,
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-control-wrapper"
                      sm={
                        Object {
                          "offset": 6,
                          "span": 18,
                        }
                      }
                      xs={
                        Object {
                          "offset": 0,
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-18 ant-col-sm-offset-6"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Button
                              block={false}
                              className="login-form-button"
                              ghost={false}
                              htmlType="submit"
                              loading={false}
                              type="primary"
                            >
                              <Wave>
                                <button
                                  className="ant-btn login-form-button ant-btn-primary"
                                  onClick={[Function]}
                                  type="submit"
                                >
                                  <span
                                    key=".0"
                                  >
                                    Save
                                  </span>
                                </button>
                              </Wave>
                            </Button>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
            </form>
          </Form>
        </div>
      </div>
    </Spin>
  </ProjectForm>
</Form(ProjectForm)>
`;
